common/variable_ticker.go

```diff
@@ -0,0 +1,267 @@
+	for {
+		// Check for control messages to update the ticker's configuration.
+		select {
+		case msg := <-t.controlChan:
+			t.handleControlMessage(msg)
+		default:
+			// No control message received, continue with the current configuration.
+		}
+
+		t.computePeriod()
+		if t.currentPeriod == 0 {
+			// Period 0 is a proxy for an infinite period, do not tick.
+			continue
+		}
+
+		// Send a tick.
+		startOfTick := time.Now()
+		select {
+		case t.tickChan <- struct{}{}:
+		case <-t.ctx.Done():
+			return
+		}
+
+		elapsed := time.Since(startOfTick)
+		sleepTime := t.currentPeriod - elapsed
+		if sleepTime < 0 {
+			// If ticks are requested less often than the configured frequency, no need to sleep.
+			continue
+		}
+
+		timer.Reset(sleepTime)
+		select {
+		case <-timer.C:
+		case <-t.ctx.Done():
+			return
+		}
+	}
+}
+
+// handleControlMessage processes control messages that update the ticker's configuration.
+func (t *VariableTicker) handleControlMessage(msg any) {
+	targetFrequency := t.targetFrequency
+	acceleration := t.acceleration
+
+	switch m := msg.(type) {
+	case *frequencyUpdate:
+		targetFrequency = m.targetFrequency
+	case *accelerationUpdate:
+		acceleration = m.acceleration
+	default:
+		// This should not be possible.
+		panic(fmt.Sprintf("invalid control message type: %T", msg))
+	}
+
+	t.anchorTime = time.Now()
+	t.anchorFrequency = t.currentFrequency
+	t.targetFrequency = targetFrequency
+	t.acceleration = acceleration
+}
+
+// computePeriod updates the current period based on configured frequency and acceleration
+func (t *VariableTicker) computePeriod() {
+	if t.currentFrequency == t.targetFrequency {
+		// shortcut, don't recompute period if the period is already correct
+		return
+	}
+
+	elapsedSinceAnchorTime := time.Since(t.anchorTime)
+
+	if t.acceleration == 0 {
+		// Acceleration zero is defined as infinite acceleration. Immediately adopt the target frequency.
+		t.currentFrequency = t.targetFrequency
+	} else if t.currentFrequency < t.targetFrequency {
+		// We are below the target frequency.
+		t.currentFrequency = t.anchorFrequency + (t.acceleration * elapsedSinceAnchorTime.Seconds())
+		if t.currentFrequency > t.targetFrequency {
+			// If we over shoot, adopt the target frequency.
+			t.currentFrequency = t.targetFrequency
+		} else {
+			// When speeding up, substitute the current frequency with the inflection frequency.
+			// This is to avoid sleeping for a very long time when starting from a low frequency.
+			t.currentFrequency = t.computeInflectionFrequency()
+		}
+	} else {
+		// We are above the target frequency.
+		t.currentFrequency = t.currentFrequency - (t.acceleration * elapsedSinceAnchorTime.Seconds())
+		if t.currentFrequency < t.targetFrequency {
+			// If we over shoot, adopt the target frequency.
+			t.currentFrequency = t.targetFrequency
+		}
+	}
+
+	if t.currentFrequency == 0 {
+		t.currentPeriod = 0
+	} else {
+		t.currentPeriod = time.Duration(float64(time.Second) / t.currentFrequency)
```

Comment:
I think you need to check for overflow here. If frequency is very small, `currentPeriod` may overflow

Reply:
I've set a minimum and maximum frequency that should protect us against overflow.

```
// Any frequency at or below this value will be interpreted as a frequency of 0 Hz. Needed to avoid overflow.
// The factor of 2 is to take care of floating point precision issues.
const MinimumFrequency = float64(time.Nanosecond/math.MaxInt64) * 2.0

// Any frequency above this value will be interpreted as a frequency of MaximumFrequency Hz. Needed to avoid overflow.
const MaximumFrequency = float64(time.Nanosecond)
```